src/assign.rs

```diff
@@ -133,55 +133,66 @@ pub trait Assign {
 ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
 ╔══════════════════════════════════════════════════════════════════════════════╗
 ║                                                                              ║
-║                                 AssignError                                  ║
-║                                ¯¯¯¯¯¯¯¯¯¯¯¯¯                                 ║
+║                                    Error                                     ║
+║                                   ¯¯¯¯¯¯¯                                    ║
 ╚══════════════════════════════════════════════════════════════════════════════╝
 ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
 */
 
+// TODO: should AssignError be deprecated?
+/// Alias for [`Error`].
+///
+/// Possible error returned from [`Assign`] implementations for
+/// [`serde_json::Value`] and
+/// [`toml::Value`](https://docs.rs/toml/0.8.14/toml/index.html).
+pub type AssignError = Error;
+
 /// Possible error returned from [`Assign`] implementations for
 /// [`serde_json::Value`] and
 /// [`toml::Value`](https://docs.rs/toml/0.8.14/toml/index.html).
 #[derive(Debug, PartialEq, Eq)]
-pub enum AssignError {
-    /// A `Token` within the `Pointer` failed to be parsed as an array index.
+pub enum Error {
+    /// A [`Token`] within the [`Pointer`] failed to be parsed as an array index.
     FailedToParseIndex {
+        /// Position (index) of the token which failed to parse as an `Index`
+        position: usize,
         /// Offset of the partial pointer starting with the invalid index.
         offset: usize,
         /// The source [`ParseIndexError`]
         source: ParseIndexError,
     },
 
-    /// target array.
+    /// A [`Token`] within the [`Pointer`] contains an [`Index`] which is out of bounds.
+    ///
+    /// The current or resulting array's length is less than the index.
     OutOfBounds {
+        /// Position (index) of the token which failed to parse as an `Index`
+        position: usize,
         /// Offset of the partial pointer starting with the invalid index.
         offset: usize,
         /// The source [`OutOfBoundsError`]
         source: OutOfBoundsError,
     },
 }
 
-impl fmt::Display for AssignError {
+impl fmt::Display for Error {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
-            Self::FailedToParseIndex { offset, .. } => {
-                write!(
-                    f,
-                    "assignment failed due to an invalid index at offset {offset}"
-                )
+            Self::FailedToParseIndex { .. } => {
+                write!(f, "assignment failed due to an invalid index")
```

Comment:
No, I don't think we should deprecate it. It actually is incredibly useful to have and a positive consequence of my blunder to opt for `offset` to begin with.

Folks can have pretty printed errors that do something like
```
/some/example/invalid/more
             ^^^^^^^^
```
without much additional effort to determine where that starts.

Reply:
I see! In that case, why did you remove it from the message? 😄

Reply:
I should have mentioned that I went through the exact same thought process as you.

I started with "ugh, I guess I add `position`, deprecate `offset`, and plan on removing it. But that path has so many breaking changes across numerous releases.

Then it dawned on me that it's actually really useful.

Reply:
Nope - I'm back on this was a mistake. I'd like to remove `offset` but I think we may be stuck with it. Or we potentially break a lot.

I just discovered [`mini-crater`](https://github.com/shepmaster/mini-crater], which may give some insight into what deleting `offset` would likely mean, at least to known dependents.

Reply:
@asmello also, I'm not sure what to do about a `Report`. For `Pointer`, we are operating on borrowed strings. `PointerBuf::parse` accepts `Into<String>` so we own the potentially only copy of that value and are not handing it back in any way - I think this may be a mistake on our part.

Reply:
i don't mean to imply `Into<String>` being the sticking point for `PointerBuf`. It's that we are likely taking ownership of the input but not handing that back with the error.

We wouldn't need to hand it back in in the case of `Pointer::parse` though.

I guess I should create an issue for this.

Reply:
Sorry, have been busy, finally getting through my backlog. I think this is actually not a big deal, because if you want to keep ownership you can always pass a reference. This will create a clone internally, but consider the alternative of taking a reference — then we'll _always_ clone, even if the caller doesn't need the string anymore (which is a fairly common case).

If we want to optimise this I think we can take a `impl Into<Cow<'_, str>>`, and return it in the error variant. This way:
- If the caller passes a `&str` and parsing succeeds, we clone internally. This is as expected.
- If the caller passes a `&str` and parsing fails, we don't clone. We return the same reference. No problem.
- If the caller passes a `String` and parsing succeeds, we just keep the string. This is the ideal scenario.
- If the caller passes a `String` and parsing fails, we return the same owned string. No problem.

I think this is worth changing, but I think it's relatively low priority.

Reply:
> Sorry, have been busy, finally getting through my backlog.

No need to apologize about that at all.

> because if you want to keep ownership you can always pass a reference. This will create a clone internally, but consider the alternative of taking a reference — then we'll always clone, even if the caller doesn't need the string anymore

So for `Pointer::parse`, I don't think we should return the string. For one, they already have the string - we literally cannot take ownership of it. Two, a reference would introduce lifetimes which would make the errors a pain. The only upside is reporting which not everyone needs but introduces additional work for everyone else to rid themselves of lifetimes with a call to `into_owned`.

`PointerBuf::parse` is different. In the case of `&str`, we have already allocated with `.into()`. In the case of an owned `String`, we have an existing allocation which we are currently dropping on the error path.

If we require the caller to keep a copy for their error path, we are causing an additional allocation on the happy path.

Right now, in #93 I have 
```rust
pub struct ParseBufError {
    pub value: String,
    pub source: ParseError,
}
```
and `ParseError` implements `From<ParseBufError>`.

I've got to run for now - I'll reply to the other when I'm in front of my laptop.

Reply:
> So for Pointer::parse, I don't think we should return the string.

Agreed, although including an owned copy there would allow us to report with the `#[source_code]` attribute without the annoying lifetime. Given this is the unhappy path I think it'd be reasonable to allocate for that.

> If we require the caller to keep a copy for their error path, we are causing an additional allocation on the happy path.

Yep, hence the suggestion to make it use a `Cow`. See my full suggestion in #96 (I think you meant to link that one?).

Reply:
> I see! In that case, why did you remove it from the message? 😄

sorry, I missed this question and I can't recall :(

Reply:
Oh!! You meant the error message!  Sorry, I should have held off on replying until I could devote enough energy to properly think through the question regarding the message. I couldn't recall what I had removed the offset from.

I removed the offset from the error message initially because I wasn't sure how much value it was adding vs making the error message more verbose and confusing. In retrospect, I'm not sure either are the case. Some may appreciate the offset in the message and I think the error message is now longer than what it was before.

I'll add them back.

Reply:
Yeah, just to clarify it's not that I think we _need_ the offset in the message, but just that the fact it's not used makes it seem like it's not useful (and if it isn't, perhaps it should be deprecated). But with miette in the picture I think it definitely _is_ useful because then it can be used for the label.